### DISCUSSION TITLE
Porting U-Boot to a custom carrier board

### DISCUSSION THREAD

**Discussion by octonout** [OE4T/meta-tegra#878] [Q&A]
Yocto version: dunfell
meta-tegra: Linux4Tegra release: R32.4.3 and JetPack release: 4.4.

Hi. I am using the [Jetson Nano Platform Adaptation and Bring-Up](https://docs.nvidia.com/jetson/archives/l4t-archived/l4t-3243/index.html#page/Tegra%2520Linux%2520Driver%2520Package%2520Development%2520Guide%2Fadaptation_and_bringup_nano.html%23wwpID0E0EO0HA) to port my Yocto image to a custom carrier board. Right now I am stuck at the part where I have to [port U-Boot](https://docs.nvidia.com/jetson/archives/l4t-archived/l4t-3243/index.html#page/Tegra%2520Linux%2520Driver%2520Package%2520Development%2520Guide%2Fadaptation_and_bringup_nano.html%23wwpID0E0EO0HA).

Currently, I do not want to change the partitioning. I am just looking to add device tree support for a device that is connected to SPI, some on-carrier-board LEDs, some buttons and a camera port that has different pin configuration compared to the 15-pin ports on the devkit board B01.

The porting guide has me a little confused. After going through the sources a bit, and also the Wiki page, it seems like what I have to do, is create a new machine conf for my custom board. But I am finding it difficult to figure out what values I need to change. Below is the jetson-nano-emmc.conf

```
#@TYPE: Machine
#@NAME: Nvidia Jetson Nano (eMMC)
#@DESCRIPTION: Nvidia Jetson Nano dev board with eMMC module

KERNEL_ARGS ?= "console=ttyS0,115200 console=tty0 fbcon=map:0 net.ifnames=0 sdhci_tegra.en_boot_part_access=1"
IMAGE_ROOTFS_ALIGNMENT ?= "1024"

NVPMODEL ?= "nvpmodel_t210_jetson-nano"

TEGRA_BOARDID ?= "3448"
TEGRA_FAB ?= "400"
TEGRA_BOARDSKU ?= "0002"
# Extracted from l4t_generate_soc_bup.sh for BOARDID=3448 and board=jetson-nano-emmc
TEGRA_BUPGEN_SPECS ?= "fab=200 fab=300 fab=400"

require conf/machine/include/tegra210.inc

KERNEL_DEVICETREE ?= "_ddot_/_ddot_/_ddot_/_ddot_/nvidia/platform/t210/porg/kernel-dts/tegra210-p3448-0002-p3449-0000-a02.dtb \
		      _ddot_/_ddot_/_ddot_/_ddot_/nvidia/platform/t210/porg/kernel-dts/tegra210-p3448-0002-p3449-0000-b00.dtb"

UBOOT_MACHINE = "p3450-0002_defconfig"

EMMC_SIZE ?= "15758000128"
EMMC_DEVSECT_SIZE ?= "512"
BOOTPART_SIZE ?= ""
BOOTPART_LIMIT ?= "10485760"
ROOTFSPART_SIZE ?= "15032385536"
ODMDATA ?= "0x94000"
EMMC_BCT ?= "P3448_A00_4GB_Micron_4GB_lpddr4_204Mhz_P987.cfg"
NVIDIA_BOARD ?= "t210ref"
NVIDIA_PRODUCT ?= "p3450-porg"
NVIDIA_BOARD_CFG ?= ""
PARTITION_LAYOUT_TEMPLATE ?= "flash_l4t_t210_emmc_p3448.xml"
```
From these values, it seems like I would have to supply a modified defconfig to `UBOOT_MACHINE` and it seems like the `NVIDIA_PRODUCT` variable might have to change? I haven't found where this variable is used though. I have already created the .dtsi files using the Pinmux Configuration spreadsheet that NVIDIA provides. And I replaced the default .dtsi files that the kernel uses to build the .dtb files that are referenced by the `KERNEL_DEVICETREE` variable.

Going through the U-Boot sources, I can see that there is a `board/nvidia/p3450-0000` directory which houses:
├── Kconfig
├── MAINTAINERS
├── Makefile
└── p3450-0000.c

It seems like I would need to create my own board files and create a patch for the U-Boot sources. Plus I would also have to add the modified defconfig referenced via `UBOOT_MACHINE` here.

Should these changes be enough? Or are any more changes required?

I would really appreciate if someone could help me out here.

**Comment by madisongh** [comment]
> It seems like I would need to create my own board files and create a patch for the U-Boot sources. Plus I would also have to add the modified defconfig referenced via UBOOT_MACHINE here.

Yep, that should do it for U-Boot.

You can ignore the `NVIDIA_PRODUCT` setting; that's not actually used anywhere.

**Reply by octonout** [reply]
Hey. I am getting back to working on it now. Thanks for the reply.

I just made a copy of the `arch/arm/dts/tegra210-p3450-0000.dts` and named it `arch/arm/dts/tegra210-<board>.dts`. Now I am editing the same file. There is this compatible string in the root node of the board dts `compatible = "nvidia,p3450-0000", "nvidia,tegra210";` From what I have learnt about the device tree, the compatible string is used to find the appropriate driver for a device in the kernel.
So, I am wondering whether I have to change this to `compatible = "<vendor>,<board>", "nvidia,tegra210";` ? Or should I leave it as is?

**Reply by madisongh** [reply]
I'm don't think there's anything in U-Boot that checks for that specific compatibility string in the device tree, but it would be a good idea to make it unique for your board, especially if you're making other changes.

**Reply by octonout** [reply]
Will try it. Thanks
